homeassistant/components/lamarzocco/config_flow.py

```diff
@@ -0,0 +1,262 @@
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except NoMachines:
+                errors["base"] = "no_machines"
+
+            if not errors:
+                if self._discovered:
+                    serials = [machine[0] for machine in self._machines]
+                    if self._discovered[CONF_MACHINE] not in serials:
+                        errors["base"] = "machine_not_found"
+                    else:
+                        self._config = data
+                        return await self.async_step_host_selection()
+
+            if not errors:
+                self._config = data
+                return await self.async_step_machine_selection()
+
+        return self.async_show_form(
+            step_id="user", data_schema=LOGIN_DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_host_selection(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Machine was discovered, only enter host."""
+        errors: dict[str, str] = {}
+        if user_input:
+            # if host is set, check if we can connect to it
+            if user_input.get(CONF_HOST):
+                lm = LaMarzoccoClient(self.hass, self._config)
+                if not await lm.check_local_connection(
+                    credentials=self._config,
+                    host=user_input[CONF_HOST],
+                    serial=self._discovered[CONF_MACHINE],
+                ):
+                    errors[CONF_HOST] = "cannot_connect"
+            if not errors:
+                return self.async_create_entry(
+                    title=self._discovered[CONF_MACHINE],
+                    data=self._config | user_input,
+                )
+        return self.async_show_form(
+            step_id="host_selection",
+            data_schema=vol.Schema({vol.Optional(CONF_HOST): cv.string}),
+            errors=errors,
+        )
+
+    async def async_step_machine_selection(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Let user select machine to connect to."""
+        errors: dict[str, str] = {}
+        if user_input:
+            serial_number = user_input[CONF_MACHINE]
+            await self.async_set_unique_id(serial_number)
+            self._abort_if_unique_id_configured()
+
+            # if host is set, check if we can connect to it
+            if user_input.get(CONF_HOST):
+                lm = LaMarzoccoClient(self.hass, self._config)
+                if not await lm.check_local_connection(
+                    credentials=self._config,
+                    host=user_input[CONF_HOST],
+                    serial=serial_number,
+                ):
+                    errors[CONF_HOST] = "cannot_connect"
+            if not errors:
+                return self.async_create_entry(
+                    title=serial_number, data=self._config | user_input
+                )
```

Comment:
This is the same code as in the step above. Please avoid duplicated code

Reply:
similar - but not the same code. If I tried to put that in a method, it would reduce readability imo.

Reply:
It's not quite the same code and forcing it into an method would not improve readability imo

Reply:
I still don't quite understand what you meant, but I moved some code to a method now.